# 软件工程与计算Ⅰ -异常处理

|                                                              |                                                              |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![image-20200821202500677](https://oss.ydjsir.com.cn/img/image-20200821202500677.png) | ![image-20200821202333221](https://oss.ydjsir.com.cn/img/image-20200821202333221.png) |

finally语句：无论结果如何都会执行的语句！哪怕是return，它都要执行！以return中返回的值为准！

注意一个问题：try代码块里面声明的变量到外面就不认了！

```java
public static void printAbstract(){
        String Hello = "";
        try(InputStreamReader inputStreamReader = new InputStreamReader(System.in);BufferedReader reader = new BufferedReader(inputStreamReader)){//把释放时持有资源的东西放到try前面的括号里面！这个是Java7新增的try-with-resources。
            Hello = reader.readLine(); }
        catch(IOException e){
            e.printStackTrace(); }
        finally {
            System.out.println("Done!"); }
        System.out.println(Hello); }
```

|                                                              |                                                              |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![image-20200821203935606](https://oss.ydjsir.com.cn/img/image-20200821203935606.png) | ![image-20200821203941895](https://oss.ydjsir.com.cn/img/image-20200821203941895.png) |

## Multiple catch blocks must be ordered from smallest to biggest

|                                                              |                                                              |                                                              |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![image-20200821204031951](https://oss.ydjsir.com.cn/img/image-20200821204031951.png) | ![image-20200821204037860](https://oss.ydjsir.com.cn/img/image-20200821204037860.png) | ![image-20200821204301653](https://oss.ydjsir.com.cn/img/image-20200821204301653.png) |

某种异常在异常继承树中位置越高，那么它被catch时的篮子就越大。可以没有catch，但是不能没有try！

函数会一层一层向外抛出异常，如果最后main也抛出异常了，整个程序就终止了。

![image-20200821204355520](https://oss.ydjsir.com.cn/img/image-20200821204355520.png)

• 异常处理不能代替简单的测试；

• 不要过分地细化异常，但可以利用异常层次结构；

• 不要只抛出RuntimeException异常，应该寻找更合适的子类或者创建⾃自己的异常类；

• 不要压制异常，在检测错误时，“苛刻”要比放任更好；

• 在出错的地方抛出一个EmptyStackException异常要比在后面抛出一个NullPointerException异常更好

• 不要羞于传递异常，早抛出，晚捕获

## 创建自己的异常

- 精心设计异常的层次结构；
- 异常类中包含足够的信息；
- 异常与错误提示；

## 异常的消失与解决方案

|                                                              |                                                              |                                                              |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![image-20200821205707395](https://oss.ydjsir.com.cn/img/image-20200821205707395.png) | ![image-20200821205804968](https://oss.ydjsir.com.cn/img/image-20200821205804968.png) | ![image-20200821205820584](https://oss.ydjsir.com.cn/img/image-20200821205820584.png) |
| 这样的话只会获得finally中产生的异常                          | 抛出try语句块中阐述的原始异常，忽略在finally语句块中产生的异常 | 把产生的异常都记录下来（addSuppressed）                      |

